### PR TITLE
Wait for cloud-provider taint to be gone before starting the netpol controller

### PR DESCRIPTION
**Pull request by manuelbuil**
#### Proposed Changes ####

To avoid race conditions (as explained in the issue) do not start the netpol controller until we are sure that the cloud-provider is ready with setting the correct addresses in the node object.

To know that it is ready, check that the taint `node.cloudprovider.kubernetes.io/uninitialized` is removed. If not, check again after a few seconds. Print a warning if after 10 minutes the tainted was not removed. In that case, something went wrong with the cloud-provider but still let network policy controller run

#### Types of Changes ####

bug fix

#### Verification ####

check issue

#### Testing ####

No

#### Linked Issues ####

https://github.com/k3s-io/k3s/issues/9071

#### User-Facing Change ####

```release-note

```

### CODE REVIEW

File: pkg/agent/netpol/netpol.go [hunk 1]
```diff
@@ -142,3 +153,33 @@ func Run(ctx context.Context, nodeConfig *config.Node) error {
...
+			// Get the node object
+			node, err := client.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+			if err != nil {
+				logrus.Errorf("Error getting the node object: %v", err)
+			}
+
+			// Check for the uninitialized taint
+			for _, taint := range node.Spec.Taints {
+				if taint.Key == cloudproviderapi.TaintExternalCloudProvider {
+					taintFound = true
```
Comment (thomasferrandiz) [inline pkg/agent/netpol/netpol.go:175]: why not return directly here?

File: pkg/agent/netpol/netpol.go [hunk 2]
```diff
@@ -55,6 +59,12 @@ func Run(ctx context.Context, nodeConfig *config.Node) error {
 		return err
 	}
 
+	// As kube-router netpol requires addresses to be available in the node object
+	// Loop until the node has ready addresses to avoid race conditions (max 1 minute).
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+	checkNodeTaint(ctx, client, nodeConfig.AgentConfig.NodeName)
```
Comment (brandond) [inline pkg/agent/netpol/netpol.go:66]: You should be able to replace both this context and the checkNodeTaints with [wait.PollImmediateUntilWithContext](https://pkg.go.dev/k8s.io/apimachinery@v0.26.0/pkg/util/wait#PollImmediateUntilWithContext), or [wait.PollImmediateWithContext](https://pkg.go.dev/k8s.io/apimachinery@v0.26.0/pkg/util/wait#PollImmediateWithContext) if you want to put a limit on how long it waits.

A good example of using this to wait for the local node's status is at https://github.com/k3s-io/k3s/blob/941119640612808cfb2d71d1c85bec74740b493a/pkg/agent/tunnel/tunnel.go#L165-L190

**Review by brandond — CHANGES_REQUESTED** [review]

File: pkg/agent/netpol/netpol.go (hunk 2 above)
Comment (brandond) [reply pkg/agent/netpol/netpol.go]: They are deprecated but is still sees wide use across the Kubernetes codebase and within our projects. The difference between the deprecated and newer methods is mostly around how they return context errors. I'd prefer to stick with the current functions until we are ready to move the entire codebase off `PollImmediateUntilWithContext`/`PollImmediateWithContext`.

Comment (manuelbuil) [reply pkg/agent/netpol/netpol.go]: Ok. But let me add a comment about it

File: pkg/agent/netpol/netpol.go [hunk 3]
```diff
@@ -55,6 +60,30 @@ func Run(ctx context.Context, nodeConfig *config.Node) error {
 		return err
 	}
 
+	// As kube-router netpol requires addresses to be available in the node object
+	// Wait until the node has ready addresses to avoid race conditions (max 1 minute).
+	// TODO: Replace with non-deprecated PollUntilContextTimeout when our and Kubernetes code migrate to it
+	wait.PollImmediateWithContext(ctx, 2*time.Second, 60*time.Second, func(ctx context.Context) (bool, error) {
+		taintFound := false
+		// Get the node object
+		node, err := client.CoreV1().Nodes().Get(context.TODO(), nodeConfig.AgentConfig.NodeName, metav1.GetOptions{})
```
Comment (brandond) [inline pkg/agent/netpol/netpol.go:69]: Use the context!
```suggestion
		node, err := client.CoreV1().Nodes().Get(ctx, nodeConfig.AgentConfig.NodeName, metav1.GetOptions{})
```

File: pkg/agent/netpol/netpol.go [hunk 4]
```diff
@@ -55,6 +60,30 @@ func Run(ctx context.Context, nodeConfig *config.Node) error {
...
+		// If there is no cloud-provider, the tain will not be there
+		for _, taint := range node.Spec.Taints {
+			if taint.Key == cloudproviderapi.TaintExternalCloudProvider {
+				taintFound = true
+			}
+		}
+		if !taintFound {
+			return true, nil
+		}
+		return false, nil
```
Comment (brandond) [inline pkg/agent/netpol/netpol.go:84]: You don't need an intermediate var, just return false as soon as you find it, and true if you don't.
```suggestion
				return false, nil
			}
		}
		return true, nil
```

File: pkg/agent/netpol/netpol.go [hunk 5]
```diff
@@ -55,6 +60,30 @@ func Run(ctx context.Context, nodeConfig *config.Node) error {
 		return err
 	}
 
+	// As kube-router netpol requires addresses to be available in the node object
+	// Wait until the node has ready addresses to avoid race conditions (max 1 minute).
+	// TODO: Replace with non-deprecated PollUntilContextTimeout when our and Kubernetes code migrate to it
+	wait.PollImmediateWithContext(ctx, 2*time.Second, 60*time.Second, func(ctx context.Context) (bool, error) {
+		taintFound := false
```
Comment (brandond) [inline pkg/agent/netpol/netpol.go:67]:
```suggestion
```

**Review by brandond — CHANGES_REQUESTED** [review]

File: pkg/agent/netpol/netpol.go (hunk 3 above)
Comment (manuelbuil) [reply pkg/agent/netpol/netpol.go:69]: :facepalm:

File: pkg/agent/netpol/netpol.go (hunk 4 above)
Comment (manuelbuil) [reply pkg/agent/netpol/netpol.go:84]: thanks Brad, you are totally right

**Review by brandond — DISMISSED** [review]

File: pkg/agent/netpol/netpol.go [hunk 6]
```diff
@@ -55,6 +60,26 @@ func Run(ctx context.Context, nodeConfig *config.Node) error {
 		return err
 	}
 
+	// As kube-router netpol requires addresses to be available in the node object
+	// Wait until the node has ready addresses to avoid race conditions (max 1 minute).
+	// TODO: Replace with non-deprecated PollUntilContextTimeout when our and Kubernetes code migrate to it
+	wait.PollImmediateWithContext(ctx, 2*time.Second, 60*time.Second, func(ctx context.Context) (bool, error) {
```
Comment (brandond) [inline pkg/agent/netpol/netpol.go:66]: Final thought - we should probably expose any errors from this?
```suggestion
	if err := wait.PollImmediateWithContext(ctx, 2*time.Second, 60*time.Second, func(ctx context.Context) (bool, error) {
```

File: pkg/agent/netpol/netpol.go [hunk 7]
```diff
@@ -55,6 +60,26 @@ func Run(ctx context.Context, nodeConfig *config.Node) error {
...
+		}
+		// Check for the uninitialized taint that should be removed by cloud-provider
+		// If there is no cloud-provider, the taint will not be there
+		for _, taint := range node.Spec.Taints {
+			if taint.Key == cloudproviderapi.TaintExternalCloudProvider {
+				return false, nil
+			}
+		}
+		return true, nil
+	})
```
Comment (brandond) [inline pkg/agent/netpol/netpol.go:81]:
```suggestion
	}); err != nil {
		return err
	}
```

File: pkg/agent/netpol/netpol.go (hunk 6 above)
Comment (manuelbuil) [reply pkg/agent/netpol/netpol.go]: yeah, I should have done that. For some reason I thought it did not return any error

**Review by brandond — APPROVED** [review]

**Review by thomasferrandiz — APPROVED** [review]

File: pkg/agent/netpol/netpol.go [hunk 8]
```diff
@@ -55,6 +60,28 @@ func Run(ctx context.Context, nodeConfig *config.Node) error {
...
 
+	// As kube-router netpol requires addresses to be available in the node object
+	// Wait until the node has ready addresses to avoid race conditions (max 1 minute).
+	// TODO: Replace with non-deprecated PollUntilContextTimeout when our and Kubernetes code migrate to it
+	if err := wait.PollImmediateWithContext(ctx, 2*time.Second, 60*time.Second, func(ctx context.Context) (bool, error) {
+		// Get the node object
+		node, err := client.CoreV1().Nodes().Get(ctx, nodeConfig.AgentConfig.NodeName, metav1.GetOptions{})
+		if err != nil {
+			logrus.Errorf("Error getting the node object: %v", err)
+			return false, err
```
Comment (brandond) [inline pkg/agent/netpol/netpol.go:71]: This probably should have been `false, nil` to avoid breaking out of the poll loop if we can't get the node.

**Pull request merged**